tests/gstreamer_mqtt/unittest_mqtt_w_helper.cc

```diff
@@ -0,0 +1,886 @@
+
+  memset (hdr.gst_caps_str, '\0', GST_MQTT_MAX_LEN_GST_CPAS_STR);
+  memcpy (hdr.gst_caps_str, caps_str,
+      MIN (strlen (caps_str), GST_MQTT_MAX_LEN_GST_CPAS_STR - 1));
+  hdr.num_mems = 1;
+  hdr.size_mems[0] = len_buf;
+
+  msg->payloadlen = GST_MQTT_LEN_MSG_HDR + len_buf;
+  msg->payload = g_try_malloc0 (msg->payloadlen);
+  ASSERT_FALSE (msg->payload == NULL);
+
+  ret = gst_element_set_state (pipeline, GST_STATE_PLAYING);
+  EXPECT_NE (ret, GST_STATE_CHANGE_FAILURE);
+
+  {
+    gboolean mapped;
+    GstBuffer *buf;
+    GstMemory *mem;
+    GstMapInfo map;
+
+    buf = gst_buffer_new_allocate (NULL, len_buf, NULL);
+    ASSERT_FALSE (buf == NULL);
+
+    mem = gst_buffer_get_all_memory (buf);
+    ASSERT_FALSE (mem == NULL);
+
+    mapped = gst_memory_map (mem, &map, GST_MAP_READ);
+    ASSERT_EQ (mapped, TRUE);
+
+    memcpy (msg->payload, &hdr, GST_MQTT_LEN_MSG_HDR);
+    memcpy (&((guint8 *) msg->payload)[GST_MQTT_LEN_MSG_HDR], map.data,
+      len_buf);
+
+    gst_memory_unmap (mem, &map);
+    gst_buffer_unref (buf);
+  }
+
+  ma_ret = std::async (std::launch::async,
+      GstMqttTestHelper::getInstance ().getCbMessageArrived (),
+      GstMqttTestHelper::getInstance ().getContext (), topic_name, 0, msg);
+  EXPECT_TRUE (ma_ret.get ());
+
+  ret = gst_element_get_state (pipeline, &cur_state, NULL, GST_CLOCK_TIME_NONE);
+  EXPECT_EQ (ret, GST_STATE_CHANGE_SUCCESS);
+  EXPECT_EQ (cur_state, GST_STATE_PLAYING);
+
+  g_free (caps_str);
+
+  /** Changing caps while the pipeline is in the GST_STATE_PLAYING state */
+  caps_str = g_strdup ("video/x-raw,width=320,height=160,format=YUY2");
+  memset (hdr.gst_caps_str, '\0', GST_MQTT_MAX_LEN_GST_CPAS_STR);
+  memcpy (hdr.gst_caps_str, caps_str,
+      MIN (strlen (caps_str), GST_MQTT_MAX_LEN_GST_CPAS_STR - 1));
+  memcpy (msg->payload, &hdr, GST_MQTT_LEN_MSG_HDR);
+
+  ma_ret = std::async (std::launch::async,
+      GstMqttTestHelper::getInstance ().getCbMessageArrived (),
+      GstMqttTestHelper::getInstance ().getContext (), topic_name, 0, msg);
+  EXPECT_TRUE (ma_ret.get ());
+
+  ret = gst_element_set_state (pipeline, GST_STATE_NULL);
+  EXPECT_NE (ret, GST_STATE_CHANGE_FAILURE);
+
+  ret = gst_element_get_state (pipeline, &cur_state, NULL, GST_CLOCK_TIME_NONE);
+  EXPECT_EQ (ret, GST_STATE_CHANGE_SUCCESS);
+  gst_object_unref (pipeline);
+
+  g_free (topic_name);
+  g_free (str_pipeline);
+  g_free (caps_str);
+}
+
+/**
+ * @brief Fail test case for mqttsrc #0 (MQTTAsync_subscribe failure case)
+ */
+TEST (testMqttSrc, src_normal_launch0_n)
+{
+  const gsize len_buf = 1024;
+  gchar *caps_str = g_strdup ("video/x-raw,width=640,height=320,format=RGB");
+  gchar *topic_name = g_strdup ("test_topic");
+  gchar *str_pipeline = g_strdup_printf (
+      "mqttsrc sub-topic=%s debug=true is-live=true num-buffers=%d ! "
+      "capsfilter caps=%s ! videoconvert ! videoscale ! fakevideosink",
+      topic_name, 1, caps_str);
+  GError *err = NULL;
+  //GstBuffer *buf;
```

Comment:
Unnecessary lines here and L680 and L778.

Reply:
@gichan-jang Removed. Thanks.